source/posix/socket.c

```diff
@@ -1367,29 +1385,62 @@ int aws_socket_shutdown_dir(struct aws_socket *socket, enum aws_channel_directio
     return AWS_OP_SUCCESS;
 }
 
+static void s_written_task(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+    (void)status;
+
+    /* this is to handle a race condition when a callback kicks off a cleanup, or the user decides
+     * to close the socket based on something they read (SSL validation failed for example).
+     * if clean_up happens when internal_refcount > 0, socket_impl is kept dangling */
+    struct aws_socket *socket = arg;
+    struct aws_allocator *allocator = socket->allocator;
+    struct posix_socket *socket_impl = socket->impl;
+    ++socket_impl->internal_refcount;
+    socket_impl->written_task_scheduled = false;
+
+    /* Notes about weird loop:
+     * 1) Use `stop_after` node so we only process the initial contents of queue.
+     *    More nodes may be added to written_queue if user calls write() from the callback.
+     *    We DO NOT want to process those new nodes until the next time the task runs.
+     *    If we simply looped until the queue was empty, we could get into a
+     *    synchronous loop of completing and writing and completing and writing...
+     *    and it would be tough for multiple sockets to share an event-loop fairly.
+     * 2) Check if queue is empty with each iteration.
+     *    If user calls close() from the callback, close() will process all
+     *    nodes in the written_queue, and the queue will be empty when the
+     *    callstack gets back to here. */
+    if (!aws_linked_list_empty(&socket_impl->written_queue)) {
+        struct aws_linked_list_node *stop_after = aws_linked_list_back(&socket_impl->written_queue);
+        do {
+            struct aws_linked_list_node *node = aws_linked_list_pop_front(&socket_impl->written_queue);
+            struct write_request *write_request = AWS_CONTAINER_OF(node, struct write_request, node);
+            size_t bytes_written = write_request->original_buffer_len - write_request->cursor_cpy.len;
+            write_request->written_fn(socket, write_request->error_code, bytes_written, write_request->write_user_data);
+            aws_mem_release(allocator, write_request);
+            if (node == stop_after) {
+                break;
+            }
+        } while (!aws_linked_list_empty(&socket_impl->written_queue));
+    }
+
+    if (--socket_impl->internal_refcount == 0) {
```

Comment:
Any reason why this doesn't use aws_refcount?

Reply:
didn't use because we didn't need atomics, but yeah it makes this cleaner.
done